data/import_data.py

```diff
@@ -16,83 +18,139 @@ async def fetch_xml(session, url):
+    return {int(x) for x in lecture.xpath(xpath)}
+
+
+def part_ids_for_course(course):
+    return set().union(*(lecture_part_ids(lecture) for lecture in course_lectures(course)))
+
+
+def is_card_valid(card):
+    return card.xpath("string(day/@ref)") != '' and card.xpath("string(time/@ref)") != ''
+
+
+def card_data(card):
+    lecture_id = card.xpath("string(lecture/@ref)")
+    lecturer_id = card.xpath("string(lecturer/@ref)")
+    classroom_id = card.xpath("string(classroom/@ref)")
+    begin = int(card.xpath("string(day/@ref)")) * 100 + int(card.xpath("string(time/@ref)")) + 6.5
+    end = begin + int(card.xpath("string(/definitions/lectures/lecture[@id=\"{}\"]/duration)".format(lecture_id)))
+    what = "C" if card.xpath("string(/definitions/lectures/lecture[@id={}]/tag)".format(lecture_id)) in ["ex",
+                                                                                                         "cv"] else "P"
+    who = card.xpath("string(/definitions/lecturers/lecturer[@id=\"{}\"]/name)".format(lecturer_id))
+    room = card.xpath("string(/definitions/classrooms/classroom[@id=\"{}\"]/name)".format(classroom_id))
+    maybe_building = card.xpath(
+        "string(/definitions/classrooms/classroom[@id=\"{}\"]/building/@ref)".format(classroom_id))
+    building = 0 if maybe_building == "" else int(maybe_building)
+    return begin, end, what, who, room, building
+
+
+def cards_for_course_and_part(course, part_id):
+    lecture_ids = [int(lecture.attrib['id']) for lecture in course_lectures(course)]
+    lecture_predicate = " or ".join(["lecture/@ref={}".format(lecture_id) for lecture_id in lecture_ids])
+    return course.xpath("/definitions/cards/card[part/@ref={}][{}]".format(part_id, lecture_predicate))
+
+
+@functools.lru_cache(1000)
+def unique_course_time_slots(course):
+    return {frozenset({card_data(card) for card in cards_for_course_and_part(course, part_id)
+                       if is_card_valid(card)})
+            for part_id in part_ids_for_course(course)}
+
+
+@functools.lru_cache(1000)
+def jsonable_time_slots(course):
+    keys = ("od", "do", "typ", "vyucujici", "mistnost", "budova")
+    time_slots_with_possible_garbage = ([dict(zip(keys, values)) for values in variant if values[0] != 0]
+                                        for variant in unique_course_time_slots(course))
+    return [time_slot_list for time_slot_list in time_slots_with_possible_garbage if len(time_slot_list) > 0]
+
+
+@functools.lru_cache(1000)
+def translate_subject_code(timetable, subject_code):
+    exceptions = {"01VAM": "VAM/B", "01VAMB": "VAM/B",
+                  "01LIP": "LIP/B", "01LIPB": "LIP/B",
+                  "17EHJE": "ZEH",
+                  "17DEZ": "DEZ", "17PTA": "DEZ",
+                  "17JBEZ": "JBEZ/", "17JABE": "JBEZ/", "17ZBJE": "JBEZ/",
+                  "17ELZ": "TCJ1", "17TCJ1": "TCJ1",
+                  "17PRE": "NRE"
+                  }
```

Comment:
Tohle je hnusné, ale vzhledem k tomu, jak jsou zprasená data v rozvrhu, je nutné udržovat seznam výjimek, protože v datovém modelu softwaru pro tvorbu rozvrhu zřejmě chybí tabulka a tak jeho uživatelé prasí sloučené hodiny přes prezentační vrstvu. 

Např. předměty 01LIP a 01LIPB se vyučují společně v "rozvrhovém předmětu" LIP/B, a u předmětů 17DEZ a 17PTA je vše přiřazeno předmětu 17DEZ ("rozvrhový předmět" DEZ), a že jde současně o výuku 17PTA je zřejmé pouze z toho, že příslušné lectures mají tag `<tag>/PTA</tag>` - v podstatě narvaný kód předmětu, co by měl být v modelu, do prezentační vrstvy přes "poznámku k hodině" v tagu (kde většinou je bývá nic nebo jen `<tag>cv</tag>` u cvičení).

Pro letní semestr bude další skupina výjimek. Příští zimní semestr mohou současné výjimky být více či méně platné v závislosti na tom, jak moc systematicky to budou prasit za rok. ;)

Zatím mě maximálně napadá to, že takovéhle věci by nakonec byly v nějakém externím souboru jako `exceptions.dat` nebo tak nějak.